car-bridge/src/messaging.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+use std::time::Duration;
+
+use async_channel::Receiver;
+use async_stream::stream;
+use async_trait::async_trait;
+use chariott_common::{
+    config::env,
+    error::{Error, ResultExt as _},
+};
+use futures::{stream::BoxStream, StreamExt as _};
+use paho_mqtt::{
+    AsyncClient, ConnectOptionsBuilder, CreateOptionsBuilder, Message, MessageBuilder,
+    MQTT_VERSION_5, QOS_2,
+};
+use tracing::info;
+
+#[async_trait]
+pub trait Subscriber {
+    type Message;
+    type Topic;
+    type Error;
+
+    /// Subscribe to a topic.
+    async fn subscribe(
+        &mut self,
+        topic: Self::Topic,
+    ) -> Result<BoxStream<'static, Self::Message>, Self::Error>;
+}
+
+#[async_trait]
+pub trait Publisher {
+    type Message;
+    type Topic;
+    type Error;
+
+    /// Publish a message.
+    async fn publish(&self, topic: Self::Topic, message: Self::Message) -> Result<(), Self::Error>;
+}
+
+pub struct MqttMessaging {
+    client: AsyncClient,
+    receiver: Receiver<Option<Message>>,
+    is_subscribed: bool,
+}
+
+impl Drop for MqttMessaging {
+    fn drop(&mut self) {
+        // Best-effort disconnect.
+        _ = self.client.disconnect(None).wait();
+    }
+}
+
+impl MqttMessaging {
+    /// Connects to the MQTT broker and starts listening on incoming messages.
+    /// If there was a persisted session before, messages may delivered before
+    /// the `connect` returns. Refer to the `Subscriber` implementation for how
+    /// to get access to the buffered messages.
+    pub async fn connect(client_id: String) -> Result<Self, Error> {
+        const BROKER_URL_ENV_NAME: &str = "BROKER_URL";
+        const DEFAULT_BROKER_URL: &str = "tcp://localhost:1883";
+        const MQTT_CLIENT_BUFFER_SIZE: usize = 200;
+
+        let host = env::<String>(BROKER_URL_ENV_NAME);
+        let host = host.as_deref().unwrap_or(DEFAULT_BROKER_URL);
```

Comment:
These belong in `main.rs` and should really be passed to `connect` as arguments.

Reply:
Done in 3db1b4234b2d95febfe8fc1f9648d139fce84059